src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -284,50 +426,82 @@ private abstract class ReceiveOperation : ReadOperation
             {
                 set => CallbackOrEvent = value;
             }
 
-            protected sealed override void InvokeCallback() =>
+            protected override void InvokeCallback() =>
                 ((Action<int, byte[], int, SocketFlags, SocketError>)CallbackOrEvent)(
                     BytesTransferred, SocketAddress, SocketAddressLen, ReceivedFlags, ErrorCode);
         }
 
-        private sealed class BufferArrayReceiveOperation : ReceiveOperation
+        private sealed class BufferMemoryReceiveOperation : ReceiveOperation
         {
-            public byte[] Buffer;
-            public int Offset;
-            public int Count;
+            public Memory<byte> Buffer;
+
+            public BufferMemoryReceiveOperation(SocketAsyncContext context) : base(context) { }
 
             protected override bool DoTryComplete(SocketAsyncContext context) =>
-                SocketPal.TryCompleteReceiveFrom(context._socket, new Span<byte>(Buffer, Offset, Count), null, Flags, SocketAddress, ref SocketAddressLen, out BytesTransferred, out ReceivedFlags, out ErrorCode);
+                SocketPal.TryCompleteReceiveFrom(context._socket, Buffer.Span, null, Flags, SocketAddress, ref SocketAddressLen, out BytesTransferred, out ReceivedFlags, out ErrorCode);
+
+            protected override void InvokeCallback()
+            {
+                var cb = (Action<int, byte[], int, SocketFlags, SocketError>)CallbackOrEvent;
+                int bt = BytesTransferred;
+                byte[] sa = SocketAddress;
+                int sal = SocketAddressLen;
+                SocketFlags rf = ReceivedFlags;
+                SocketError ec = ErrorCode;
+
+                AssociatedContext.ReturnOperation(this);
+
+                cb(bt, sa, sal, rf, ec);
+            }
         }
 
         private sealed class BufferListReceiveOperation : ReceiveOperation
         {
             public IList<ArraySegment<byte>> Buffers;
 
+            public BufferListReceiveOperation(SocketAsyncContext context) : base(context) { }
+
             protected override bool DoTryComplete(SocketAsyncContext context) =>
                 SocketPal.TryCompleteReceiveFrom(context._socket, default(Span<byte>), Buffers, Flags, SocketAddress, ref SocketAddressLen, out BytesTransferred, out ReceivedFlags, out ErrorCode);
+
+            protected override void InvokeCallback()
+            {
+                var cb = (Action<int, byte[], int, SocketFlags, SocketError>)CallbackOrEvent;
+                int bt = BytesTransferred;
+                byte[] sa = SocketAddress;
+                int sal = SocketAddressLen;
+                SocketFlags rf = ReceivedFlags;
+                SocketError ec = ErrorCode;
+
+                AssociatedContext.ReturnOperation(this);
+
+                cb(bt, sa, sal, rf, ec);
+            }
         }
 
         private sealed unsafe class BufferPtrReceiveOperation : ReceiveOperation
```

Comment:
Just to make sure I understand:  We're not caching these, and these are used for synchronous Span<byte> calls, right?  I think that seems fine, but it's a little weird.

Reply:
Right. I only cached things that seemed to be places we cared about high performance.  If we're using this, we're making a synchronous call on a non-blocking socket, we're allocating a ManualResetEventSlim, we're blocking a thread while waiting for the asynchronous operation to complete, etc.  Caching here didn't seem to be a good tradeoff.  But I can add it if you think it's warranted.

Reply:
Nope, I'm happy as is, just want to make sure I understand.

Reply:
So synchronous calls on a non blocking socket doesn’t get the benefits?

Reply:
@davidfowl Yep.  More specifically, sync calls on a nonblocking socket that don't complete immediately (i.e. return EWOULDBLOCK)

Reply:
Wait, to be clear, we're talking about synchronous calls on a socket that the user has left with the default of Blocking but that under the covers we've transitioned to actually be non-blocking because an asynchronous operation was previously made on it.  If the developer has explicitly put the socket into non-blocking mode (e.g. `s.Blocking = false;`) then we don't come near this code, right @geoffkizer?

Reply:
@stephentoub Yes, that's correct.  The socket has become non-blocking *in the OS* because of a previous async operation.  The socket is still in non-blocking mode from the user point of view, i.e. s.Blocking = false.

Reply:
> The socket is still in non-blocking mode from the user point of view, i.e. s.Blocking = false.

Still confused... don't you mean "The Socket is still in blocking mode from the user point of view, i.e. s.Blocking = true"?

Reply:
Ugh, yes, correct.

The Socket is still in blocking mode from the user point of view, i.e. s.Blocking = true.

Reply:
Ah, that's fine then.